### PR TITLE
Deactivate previous env before activating new one

### PR DESCRIPTION
**Pull request by haampie**
Currently on develop you can run `spack env activate` multiple times to switch
between environments, but they leave traces, even though Spack only supports
one active environment at a time.

Currently:

```
$ spack env create a
$ spack env create b
$ spack env activate -p a
[a] $ spack env activate -p b
[b] [a] $ spack env activate -p a
[a] [b] [a] $ echo $MANPATH | tr ":" "\n"
/path/to/environments/a/.spack-env/view/share/man
/path/to/environments/a/.spack-env/view/man
/path/to/environments/b/.spack-env/view/share/man
/path/to/environments/b/.spack-env/view/man
```

This PR fixes that by deactivating the previous environment before activating a new one (basically merging env modifications for deactiviate + activate and applying them after the shell "headers" for activation / deactivation)

```
$ spack env activate -p a
[a] $ spack env activate -p b
[b] $ spack env activate -p a
[a] $ echo $MANPATH | tr ":" "\n"
/path/to/environments/a/.spack-env/view/share/man
/path/to/environments/a/.spack-env/view/man
```

With this PR shell modifications work like this:

1. Activating a new environment with no prior environment activated:

```console
$ spack env activate --sh -p a
export SPACK_ENV=/environments/a;
alias despacktivate='spack env deactivate';
if [ -z ${SPACK_OLD_PS1+x} ]; then
    if [ -z ${PS1+x} ]; then
        PS1='$$$$';
    fi;
    export SPACK_OLD_PS1="${PS1}";
fi;
export PS1="[a]  ${PS1}";
export CMAKE_PREFIX_PATH=/environments/a/.spack-env/view;
export MANPATH=/environments/a/.spack-env/view/share/man:/environments/a/.spack-env/view/man;
export PATH='/environments/a/.spack-env/view/bin:/usr/local/go/bin:~/.cargo/bin:/home/harmen/bin:/home/harmen/spack/bin:/home/harmen/clingo/.spack-env/view/bin:/home/harmen/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin';
export ACLOCAL_PATH=/environments/a/.spack-env/view/share/aclocal;
export PKG_CONFIG_PATH=/environments/a/.spack-env/view/share/pkgconfig:/environments/a/.spack-env/view/lib64/pkgconfig:/environments/a/.spack-env/view/lib/pkgconfig;
```

2. "Reactivating" the environment (usually a no-op, but it will also modify the environment variables in case you somehow unset them):

```console
$ spack env activate -p a
[a] $ spack env activate --sh a
if [ ! -z ${SPACK_ENV+x} ]; then
unset SPACK_ENV; export SPACK_ENV;
fi;
unalias despacktivate;
if [ ! -z ${SPACK_OLD_PS1+x} ]; then
    if [ "$SPACK_OLD_PS1" = '$$$$' ]; then
        unset PS1; export PS1;
    else
        export PS1="$SPACK_OLD_PS1";
    fi;
    unset SPACK_OLD_PS1; export SPACK_OLD_PS1;
fi;
export SPACK_ENV=/environments/a;
alias despacktivate='spack env deactivate';
if [ -z ${SPACK_OLD_PS1+x} ]; then
    if [ -z ${PS1+x} ]; then
        PS1='$$$$';
    fi;
    export SPACK_OLD_PS1="${PS1}";
fi;
export PS1="[a]  ${PS1}";
```

3. Switching environments from `a` to `b` (undoes env a, applies b):

```console
if [ ! -z ${SPACK_ENV+x} ]; then
unset SPACK_ENV; export SPACK_ENV;
fi;
unalias despacktivate;
if [ ! -z ${SPACK_OLD_PS1+x} ]; then
    if [ "$SPACK_OLD_PS1" = '$$$$' ]; then
        unset PS1; export PS1;
    else
        export PS1="$SPACK_OLD_PS1";
    fi;
    unset SPACK_OLD_PS1; export SPACK_OLD_PS1;
fi;
export SPACK_ENV=/environments/b;
alias despacktivate='spack env deactivate';
if [ -z ${SPACK_OLD_PS1+x} ]; then
    if [ -z ${PS1+x} ]; then
        PS1='$$$$';
    fi;
    export SPACK_OLD_PS1="${PS1}";
fi;
export PS1="[b]  ${PS1}";
export PKG_CONFIG_PATH=/environments/b/.spack-env/view/share/pkgconfig:/environments/b/.spack-env/view/lib64/pkgconfig:/environments/b/.spack-env/view/lib/pkgconfig;
export LD_LIBRARY_PATH=/environments/b/.spack-env/view/lib64:/environments/b/.spack-env/view/lib;
export PATH='/environments/b/.spack-env/view/bin:/usr/local/go/bin:~/.cargo/bin:/home/harmen/bin:/home/harmen/spack/bin:/home/harmen/clingo/.spack-env/view/bin:/home/harmen/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin';
export ACLOCAL_PATH=/environments/b/.spack-env/view/share/aclocal;
export MANPATH=/environments/b/.spack-env/view/share/man:/environments/b/.spack-env/view/man;
export CMAKE_PREFIX_PATH=/environments/b/.spack-env/view;
```

### CODE REVIEW

File: lib/spack/spack/cmd/env.py [hunk 1]
```diff
@@ -89,26 +91,39 @@ def env_activate(args):
...
 
-    if spack_env == os.environ.get('SPACK_ENV'):
-        tty.debug("Environment %s is already active" % args.activate_env)
-        return
+    # Shell commands
+    cmds = ''
+
+    # Deactivate the current environment first.  We don't properly support
+    # stacked environments at the moment.
+    if ev.active_environment() is not None:
```
Comment (haampie) [inline lib/spack/spack/cmd/env.py]: The only potential issue here is that `-e` can override currently active env. We should probably just error out for stuff like this

```
spack -e ... env active ...
spack -e ... env deactivate
```

or not @becker33?

Comment (haampie) [reply lib/spack/spack/cmd/env.py]: This is resolved in #25625

**Review by tldahlgren — DISMISSED** [review]
Tested the basic commands and they worked for me!

**Comment by alalazo** [comment]
@spackbot run pipeline

**Comment by alalazo** [comment]
@scheibelp I think this solves #25314 since:
```console
$ spack env activate a
$ spack env activate a # This should deactivate a and reactivate it
```
is supposed to be a valid operation after this PR (I didn't actually check that, but that's my understanding reading the description).

File: lib/spack/spack/cmd/env.py [hunk 2]
```diff
@@ -115,42 +116,46 @@ def env_activate(args):
...
     else:
         tty.die("No such environment: '%s'" % env_name_or_dir)
 
     env_prompt = '[%s]' % short_name
 
-    if spack_env == os.environ.get('SPACK_ENV'):
-        tty.debug("Environment is already active")
-        return
+    # Deactivate the current environment first.  We don't properly support
+    # stacked environments at the moment.
```
Comment (scheibelp) [inline lib/spack/spack/cmd/env.py]: There should probably be a link to somewhere that explains what "stacked environments" are, or this should be replaced with a description of the effects like:

```
# Deactivate current env first. We don't support combining environments together.
```

File: share/spack/qa/test-framework.sh [hunk 3]
```diff
@@ -132,6 +132,33 @@ contains() {
...
+does_not_contain() {
+    string="$1"
+    shift
+
+    printf "'%s' output does not contain '$string' ... " "$*"
+    output=$("$@" 2>&1)
+    err="$?"
+
+    if [ "${output#*$string}" = "${output}" ]; then
+        pass
```
Comment (scheibelp) [inline share/spack/qa/test-framework.sh:151]: If the command fails but has no output, should this still fail (i.e. should we always fail if the return code is nonzero)?

File: share/spack/qa/setup-env-test.fish [hunk 4]
```diff
@@ -88,7 +88,7 @@ end
 function spt_succeeds
     printf "'$argv' succeeds ... "
 
-    set -l output (eval $argv 2>&1)
+    set -l output ($argv 2>&1)
```
Comment (scheibelp) [inline share/spack/qa/setup-env-test.fish:91]: (question) I don't know anything about fish: are `(eval $argv)` and `($argv)` equivalent? Or was this required due to other changes?

File: lib/spack/spack/cmd/env.py [hunk 5]
```diff
@@ -181,10 +186,10 @@ def env_deactivate(args):
 
     # Error out when -e, -E, -D flags are given, cause they are ambiguous.
     if args.env or args.no_env or args.env_dir:
-        tty.die('Calling spack env deactivate with --env, --env-dir and --no-env '
+        tty.die('Calling spack env activate with --env, --env-dir and --no-env '
```
Comment (scheibelp) [inline lib/spack/spack/cmd/env.py]: I'm confused why "deactivate" was replaced with "activate" here: we are in the `env_deactivate` function.

**Comment by haampie** [comment]
Even better: it's reloading the effective changes, so for instance:

```
spack env create a
spack -e a add py-flake8
spack -e a install
spack env activate a
unset CMAKE_PREFIX_PATH
spack env activate --sh a
```

outputs

```
if [ ! -z ${SPACK_ENV+x} ]; then
unset SPACK_ENV; export SPACK_ENV;
fi;
unalias despacktivate;
if [ ! -z ${SPACK_OLD_PS1+x} ]; then
    if [ "$SPACK_OLD_PS1" = '$$$$' ]; then
        unset PS1; export PS1;
    else
        export PS1="$SPACK_OLD_PS1";
    fi;
    unset SPACK_OLD_PS1; export SPACK_OLD_PS1;
fi;
export SPACK_ENV=/home/user/spack/var/spack/environments/a;
alias despacktivate='spack env deactivate';
export CMAKE_PREFIX_PATH=/home/user/spack/var/spack/environments/a/.spack-env/view;
```

this is also useful when you first activate, then add & install new specs, and you have to get their `setup_run_environment` variables, you'll have to execute `spack env activate xyz` again. Maybe we should actually make `spack env activate` without name/path a thing, which would just re-activate the current env.

File: lib/spack/spack/cmd/env.py [hunk 6]
```diff
@@ -115,42 +116,46 @@ def env_activate(args):
...
-        tty.debug("Environment is already active")
-        return
+    # Deactivate the current environment first.  We don't properly support
+    # stacked environments at the moment.
+    if ev.active_environment() is None:
+        cmds = ''
+        env_mods = EnvironmentModifications()
+    else:
+        cmds = spack.environment.shell.deactivate_header(shell=args.shell)
+        env_mods = spack.environment.shell.deactivate()
```
Comment (scheibelp) [inline lib/spack/spack/cmd/env.py:144]: (minor) If the user changes the environment after activating it (e.g. removes packages from the env), it is possible that the deactivation could miss undoing some changes. This is no worse than what happens if the user has to manually deactivate/activate the envs themselves. I think it's worth mentioning somewhere though (maybe in `environment.shell.deactivate`?)

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a few questions and maybe one or two minor changes to suggest, but overall this looks great.

File: lib/spack/spack/cmd/env.py (hunk 2 above)
Comment (haampie) [reply lib/spack/spack/cmd/env.py]: Yeah, I think I got stacked environments as a concept from Julia

Comment (tgamblin) [reply lib/spack/spack/cmd/env.py]: Julia can pull this off because they set their ABI requirement upstream -- you have to build in Yggdrasil to be in Julia in the first place, and it has one ABI it's targeting.  If we did this in Spack I think we would need stronger runtime checks about which envs can even be composed together.  For Spack-built stuff we RPATH everything so if we have DAG consistency, but I think in envs we likely want to do more than that...

File: share/spack/qa/setup-env-test.fish (hunk 4 above)
Comment (haampie) [reply share/spack/qa/setup-env-test.fish:91]: it was required for something, but you'd have to ask my brain from 2 months ago :grimacing:

File: lib/spack/spack/cmd/env.py (hunk 5 above)
Comment (haampie) [reply lib/spack/spack/cmd/env.py]: rebase issue I think

File: share/spack/qa/setup-env-test.fish (hunk 4 above)
Comment (haampie) [reply share/spack/qa/setup-env-test.fish:91]: It's probably that `spt_contains 'spack_test_2_env' 'fish' '-c' 'echo $PATH'` would otherwise evaluate $PATH too early

Comment (haampie) [reply share/spack/qa/setup-env-test.fish:91]: but generally, it seems eval is not required here, so not sure why it was there in the first place

File: share/spack/qa/test-framework.sh (hunk 3 above)
Comment (haampie) [reply share/spack/qa/test-framework.sh:151]: I think we do this on purpose so we can test error messages too (see also `2>&1` a few lines up).

Comment (scheibelp) [reply share/spack/qa/test-framework.sh:151]: > I think we do this on purpose 

I think you are referring to the implementation of `contains`. Note it is "easier" to catch failures in that function: all sorts of failures that lead to an early termination will also lead to the expected output not being generated (and we therefore detect a problem); in this case, early terminations (e.g. including those with no output) may be failures but in would not be detected as such.

In other words, I think a reasonable definition of `does_not_contain` confirms two things:

* The command succeeds
* The command output does not contain the specified string

Right now I think this function is only doing the latter. There may be cases where a user wants to confirm that a command fails and does not generate some output, but in that case this function could use a flag to enable that behavior.

Comment (haampie) [reply share/spack/qa/test-framework.sh:151]: agreed, fixed that.

**Comment by tldahlgren** [comment]
Haven't looked at the code yet to see what is going on but I'm seeing the following behavior:

```
$ spack env create a
$ spack env activate a
$ printenv > act-a.txt
$ spack env activate a
SPACK_OLD_PROMPT: Undefined variable.
$ printenv > act2-a.txt
$ diff act-a.txt act2-a.txt
< SPACK_LD_LIBRARY_PATH=/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib:/usr/tce/packages/intel/intel-19.0.4/lib/intel64
< SPACK_ENV=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a
---
> SPACK_LD_LIBRARY_PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib64:/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib:/usr/tce/packages/intel/intel-19.0.4/lib/intel64
$ spack env deactivate
==> Error: No environment is currently active.
$ printenv > deact-a.txt
$ diff deact-a.txt act2-a.txt   # Note no differences
$ spack env remove a
$ printenv > rem-a.txt
$ diff act2-a.txt rem-a.txt   # Note no differences
```

And environment variables still have paths to the view.

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Marking as request changes due to the strange behavior I am seeing as noted in my prior comment.

**Comment by scheibelp** [comment]
@tldahlgren thanks for checking that - could you also mention which shell you were in (e.g. bash, csh, etc.)?

**Comment by tldahlgren** [comment]
> @tldahlgren thanks for checking that - could you also mention which shell you were in (e.g. bash, csh, etc.)?

Sure. `tcsh`.

**Comment by scheibelp** [comment]
>> could you also mention which shell you were in (e.g. bash, csh, etc.)?

> tcsh

Out of curiosity, are you able to replicate this issue ~(a) with a fresh environment~(EDIT: your example is with a fresh env) and (b) with bash? I'm curious about the extent of the issues.

I'm currently trying to replicate this myself.

**Comment by scheibelp** [comment]
Oh also @tldahlgren do you get different results if you use `--csh` with `spack env activate` (like `spack env activate --csh a`)?

**Comment by tldahlgren** [comment]
> > > could you also mention which shell you were in (e.g. bash, csh, etc.)?
> 
> > tcsh
> 
> Out of curiosity, are you able to replicate this issue ~(a) with a fresh environment~(EDIT: your example is with a fresh env) and (b) with bash? I'm curious about the extent of the issues.
> 
> I'm currently trying to replicate this myself.

I'm trying to run in a bash shell but am getting errors with `setup-env.sh`.  Will keep trying.

**Comment by tldahlgren** [comment]
> > > > could you also mention which shell you were in (e.g. bash, csh, etc.)?
> > 
> > 
> > > tcsh
> > 
> > 
> > Out of curiosity, are you able to replicate this issue ~(a) with a fresh environment~(EDIT: your example is with a fresh env) and (b) with bash? I'm curious about the extent of the issues.
> > I'm currently trying to replicate this myself.
> 
> I'm trying to run in a bash shell but am getting errors with `setup-env.sh`. Will keep trying.

Based on a bash script that runs the sequence the results are definitely different as shown below:

```
$ spack env create a
==> Updating view at /usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view
==> Created environment 'a' in /usr/WS1/dahlgren/releases/spack/var/spack/environments/a
==> You can activate this environment with:
==>   spack env activate a
$ spack env activate a
$ printenv >& act-a.txt
$ spack env activate a
$ printenv >& act2-a.txt
$ diff act-a.txt act2-a.txt
40c40
< SPACK_LD_LIBRARY_PATH=/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib:/usr/tce/packages/intel/intel-19.0.4/lib/intel64
---
> SPACK_LD_LIBRARY_PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib64:/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib:/usr/tce/packages/intel/intel-19.0.4/lib/intel64
$ spack env deactivate
$ printenv >& deact-a.txt
$ diff deact-a.txt act2-a.txt
0a1
> SPACK_ENV=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a
3c4
< MANPATH=/usr/tce/packages/git/git-2.29.1/share/man:/usr/tce/man:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/man:/usr/tce/packages/intel/intel-19.0.4/man/common:/usr/share/lmod/lmod/share/man:/usr/tce/packages/dotkit/dotkit/man:/usr/man:/usr/share/man:/usr/local/man:/usr/X11R6/man:/usr/lib64/mvapich/default/man
---
> MANPATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/share/man:/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/man:/usr/tce/packages/git/git-2.29.1/share/man:/usr/tce/man:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/man:/usr/tce/packages/intel/intel-19.0.4/man/common:/usr/share/lmod/lmod/share/man:/usr/tce/packages/dotkit/dotkit/man:/usr/man:/usr/share/man:/usr/local/man:/usr/X11R6/man:/usr/lib64/mvapich/default/man
31c32
< LD_LIBRARY_PATH=/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib:/usr/tce/packages/intel/intel-19.0.4/lib/intel64
---
> LD_LIBRARY_PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib64:/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib:/usr/tce/packages/intel/intel-19.0.4/lib/intel64
46c47
< PATH=/usr/WS1/dahlgren/releases/spack/bin:/usr/tce/packages/git-lfs/git-lfs-2.9.2/bin:/usr/tce/packages/git/git-2.29.1/bin:/usr/tce/bin:/usr/tce/packages/texlive/texlive-2016/2016/bin/x86_64-linux:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/bin:/usr/tce/packages/intel/intel-19.0.4/bin:/usr/lib64/qt-3.3/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:.
---
> PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/bin:/usr/WS1/dahlgren/releases/spack/bin:/usr/tce/packages/git-lfs/git-lfs-2.9.2/bin:/usr/tce/packages/git/git-2.29.1/bin:/usr/tce/bin:/usr/tce/packages/texlive/texlive-2016/2016/bin/x86_64-linux:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/bin:/usr/tce/packages/intel/intel-19.0.4/bin:/usr/lib64/qt-3.3/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:.
86c87
< PKG_CONFIG_PATH=
---
> PKG_CONFIG_PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/share/pkgconfig:/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib64/pkgconfig:/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib/pkgconfig
91c92
< ACLOCAL_PATH=
---
> ACLOCAL_PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/share/aclocal
95c96
< CMAKE_PREFIX_PATH=
---
> CMAKE_PREFIX_PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view
$ spack env remove -y a
==> Successfully removed environment 'a'
$ printenv >& rem-a.txt
$ diff act2-a.txt rem-a.txt
1d0
< SPACK_ENV=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a
4c3
< MANPATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/share/man:/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/man:/usr/tce/packages/git/git-2.29.1/share/man:/usr/tce/man:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/man:/usr/tce/packages/intel/intel-19.0.4/man/common:/usr/share/lmod/lmod/share/man:/usr/tce/packages/dotkit/dotkit/man:/usr/man:/usr/share/man:/usr/local/man:/usr/X11R6/man:/usr/lib64/mvapich/default/man
---
> MANPATH=/usr/tce/packages/git/git-2.29.1/share/man:/usr/tce/man:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/man:/usr/tce/packages/intel/intel-19.0.4/man/common:/usr/share/lmod/lmod/share/man:/usr/tce/packages/dotkit/dotkit/man:/usr/man:/usr/share/man:/usr/local/man:/usr/X11R6/man:/usr/lib64/mvapich/default/man
32c31
< LD_LIBRARY_PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib64:/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib:/usr/tce/packages/intel/intel-19.0.4/lib/intel64
---
> LD_LIBRARY_PATH=/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib:/usr/tce/packages/intel/intel-19.0.4/lib/intel64
40c39
< SPACK_LD_LIBRARY_PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib64:/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib:/usr/tce/packages/intel/intel-19.0.4/lib/intel64
---
> SPACK_LD_LIBRARY_PATH=/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib:/usr/tce/packages/intel/intel-19.0.4/lib/intel64
47c46
< PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/bin:/usr/WS1/dahlgren/releases/spack/bin:/usr/tce/packages/git-lfs/git-lfs-2.9.2/bin:/usr/tce/packages/git/git-2.29.1/bin:/usr/tce/bin:/usr/tce/packages/texlive/texlive-2016/2016/bin/x86_64-linux:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/bin:/usr/tce/packages/intel/intel-19.0.4/bin:/usr/lib64/qt-3.3/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:.
---
> PATH=/usr/WS1/dahlgren/releases/spack/bin:/usr/tce/packages/git-lfs/git-lfs-2.9.2/bin:/usr/tce/packages/git/git-2.29.1/bin:/usr/tce/bin:/usr/tce/packages/texlive/texlive-2016/2016/bin/x86_64-linux:/usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/bin:/usr/tce/packages/intel/intel-19.0.4/bin:/usr/lib64/qt-3.3/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:.
87c86
< PKG_CONFIG_PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/share/pkgconfig:/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib64/pkgconfig:/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/lib/pkgconfig
---
> PKG_CONFIG_PATH=
92c91
< ACLOCAL_PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view/share/aclocal
---
> ACLOCAL_PATH=
96c95
< CMAKE_PREFIX_PATH=/usr/WS1/dahlgren/releases/spack/var/spack/environments/a/.spack-env/view
---
> CMAKE_PREFIX_PATH=
```

**Comment by scheibelp** [comment]
As for your initial issue in https://github.com/spack/spack/pull/25409#issuecomment-953132990:

```
$ spack env activate a
SPACK_OLD_PROMPT: Undefined variable.
```

There seems to be a problem with our generated csh commands where we don't enclose them in ``` `` ``` and they therefore are evaluated, for example:

```
if ( $?SPACK_OLD_PROMPT ) set prompt="$SPACK_OLD_PROMPT" && unsetenv SPACK_OLD_PROMPT ;
```

Appears in `spack env activate --csh <name-of-env>`, but this fails in my env when `$SPACK_OLD_PROMPT` isn't defined because it tries to evaluate the variable inside the statement (even if the statement itself isn't evaluated). So the following appears to get desirable behavior:

```
if ( $?SPACK_OLD_PROMPT ) `set prompt="$SPACK_OLD_PROMPT" && unsetenv SPACK_OLD_PROMPT` ;
```

I likewise had to make a similar edit to `pathadd.csh` to get `setup-env.csh` to work:

```
diff --git a/share/spack/csh/pathadd.csh b/share/spack/csh/pathadd.csh
index ec394e8990..7f3424518a 100644
--- a/share/spack/csh/pathadd.csh
+++ b/share/spack/csh/pathadd.csh
@@ -22,7 +22,7 @@ endif
 set _pa_old_value = ""
 eval set _pa_set = '$?'$_pa_varname
 if ($_pa_set == 1) then
-    eval set _pa_old_value='$'$_pa_varname
+    eval `set _pa_old_value = '$'$_pa_varname`
 endif
 
 # Do the actual prepending here, if it is a dir and not already in the path
```

**Comment by haampie** [comment]
I didn't read closely but a difference in `SPACK_LD_LIBRARY_PATH` is expected as it stores the original LD_LIBRARY_PATH before the command is run, so to see it unchanged due to re-activation, you'd have to run in a third time:

```
$ spack env activate x # SPACK_LD_LIBRARYPATH <- LD_LIBRARY_PATH from before env activation
$ spack env activate x # SPACK_LD_LIBRARYPATH <- LD_LIBRARY_PATH from after first env activation
$ env > a
$ spack env activate x # SPACK_LD_LIBRARYPATH <- LD_LIBRARY_PATH from after second env activation
$ env > b
$ diff a b
[nothing]
```

Otherwise behavior looks alright for bash, no?

Did this PR cause that problem with csh? I don't think I've touched it, but I might have messed something up :D

**Comment by scheibelp** [comment]
Regarding the third experiment in https://github.com/spack/spack/pull/25409#issuecomment-953315753:

```
$ spack env remove -y a
==> Successfully removed environment 'a'
$ printenv >& rem-a.txt
$ diff act2-a.txt rem-a.txt
```

Those results look good to me (e.g. `CMAKE_PREFIX_PATH` isn't set when the env is deactivated), was there a reason you included that?

Likewise the results of the second experiment look good:

```
$ spack env deactivate
$ printenv >& deact-a.txt
$ diff deact-a.txt act2-a.txt
```

**Comment by tldahlgren** [comment]
> Regarding the third experiment in [#25409 (comment)](https://github.com/spack/spack/pull/25409#issuecomment-953315753):
> 
> ```
> $ spack env remove -y a
> ==> Successfully removed environment 'a'
> $ printenv >& rem-a.txt
> $ diff act2-a.txt rem-a.txt
> ```
> 
> Those results look good to me (e.g. `CMAKE_PREFIX_PATH` isn't set when the env is deactivated), was there a reason you included that?

Nope.  It's WYSIWYG from me running a bash script with all the same commands I did from the (`tcsh`) shell.

> 
> Likewise the results of the second experiment look good:
> 
> ```
> $ spack env deactivate
> $ printenv >& deact-a.txt
> $ diff deact-a.txt act2-a.txt
> ```

👍

**Comment by tldahlgren** [comment]
> Oh also @tldahlgren do you get different results if you use `--csh` with `spack env activate` (like `spack env activate --csh a`)?

There's more output from `spack env activate --csh a` (i.e., `setenv`s and `alias`s) but no difference for `diff act-a.txt act2-a.txt`.  Otherwise, the output matched that I reported in https://github.com/spack/spack/pull/25409#issuecomment-953132990.

**Comment by scheibelp** [comment]
> Otherwise behavior looks alright for bash, no?

To confirm: yes it looks correct (based on https://github.com/spack/spack/pull/25409#issuecomment-953354001 and https://github.com/spack/spack/pull/25409#issuecomment-953365637

> Did this PR cause that problem with csh? I don't think I've touched it, but I might have messed something up :D

I don't think this PR introduced a *new* issue with csh, but may have revealed an existing one (e.g. https://github.com/spack/spack/pull/25409#issuecomment-953332271), which I think we would want to fix

**Comment by haampie** [comment]
It's an existing problem indeed

```
spack (develop)$ git rev-parse HEAD
890095e876a85d91ae7359f9120082c4e2f6a249 

spack (develop)$ env -i PATH=$PATH SPACK_ROOT=$PWD csh

% source share/spack/setup-env.csh 

% spack env create example
==> Updating view at /Users/user/spack/var/spack/environments/example/.spack-env/view
==> Created environment 'example' in /Users/user/spack/var/spack/environments/example
==> You can activate this environment with:
==>   spack env activate example

% spack env activate example

% spack env deactivate 
SPACK_OLD_PROMPT: Undefined variable.
```

So it should not be a blocker for this PR, but it would be nice to have it fixed in another PR

**Review by scheibelp — APPROVED** [review]

**Comment by scheibelp** [comment]
@tldahlgren based on the investigation in https://github.com/spack/spack/pull/25409#issuecomment-953382430, I'm inclined to agree that addressing csh issues can be handled in a separate PR. Does that seem agreeable to you?

**Review by tldahlgren — APPROVED** [review]
